// fortune/dist/lib/common/errors.js
'use strict';

var _Set = require('babel-runtime/core-js/set')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _errorClass = require('error-class');

var _errorClass2 = _interopRequireDefault(_errorClass);

var BadRequestError = (0, _errorClass2['default'])('BadRequestError');
exports.BadRequestError = BadRequestError;
var UnauthorizedError = (0, _errorClass2['default'])('UnauthorizedError');
exports.UnauthorizedError = UnauthorizedError;
var ForbiddenError = (0, _errorClass2['default'])('ForbiddenError');
exports.ForbiddenError = ForbiddenError;
var NotFoundError = (0, _errorClass2['default'])('NotFoundError');
exports.NotFoundError = NotFoundError;
var MethodError = (0, _errorClass2['default'])('MethodError');
exports.MethodError = MethodError;
var NotAcceptableError = (0, _errorClass2['default'])('NotAcceptableError');
exports.NotAcceptableError = NotAcceptableError;
var ConflictError = (0, _errorClass2['default'])('ConflictError');
exports.ConflictError = ConflictError;
var UnsupportedError = (0, _errorClass2['default'])('UnsupportedError');

exports.UnsupportedError = UnsupportedError;
// White-list native error types.
var nativeErrors = new _Set([Error, TypeError, ReferenceError, RangeError, SyntaxError, EvalError, URIError]);
exports.nativeErrors = nativeErrors;
